sketch/tensor_full.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "immintrin.h" // for AVX
+#include "nmmintrin.h" // for SSE4.2
+#include "sequence/alphabets.hpp"
+#include "sketch//sketch_base.hpp"
+#include "util/multivec.hpp"
+#include "util/timer.hpp"
+#include "util/utils.hpp"
+
+#include <cassert>
+#include <cmath>
+#include <iostream>
+#include <random>
+
+namespace ts { // ts = Tensor Sketch
+
+/**
+ * Computes tensor sketches for a given sequence as described in
+ * https://www.biorxiv.org/content/10.1101/2020.11.13.381814v1
+ * @tparam seq_type the type of elements in the sequences to be sketched.
+ */
+template <class seq_type>
+class FullTensor : public SketchBase<std::vector<double>, false> {
+  public:
+    // Tensor sketch output should be transformed if the command line flag is set.
+    constexpr static bool transform_sketches = false;
+
+    /**
+     * @param alphabet_size the number of elements in the alphabet S over which sequences are
+     * defined (e.g. 4 for DNA)
+     * @param normalize when true the counts will be normalized to relative frequencies with sum 1.
+     */
+    FullTensor(seq_type alphabet_size,
+               int32_t sequence_len,
+               const std::string &name = "Tensor",
+               bool normalize = true)
+        : SketchBase<std::vector<double>, false>(name),
+          alphabet_size(alphabet_size),
+          sequence_len(sequence_len),
+          normalize(normalize) {}
+
+    void init() {}
+
+    /**
+     * Computes the sketch of the given sequence.
+     * @param seq the sequence to be sketched
+     * @return an array of size alphabet_size^sequence_len containing the sequence's sketch
+     */
+    std::vector<double> compute(const std::vector<seq_type> &seq) {
+        // ts[i] contains the counts for subsequences of length i.
+        Vec2D<double> ts(sequence_len + 1);
+        {
+            uint32_t length = 1;
+            for (auto &t : ts) {
+                t.resize(length);
+                length *= alphabet_size;
+            }
+        }
+
+        // The base case is the one empty sequence.
+        ts[0][0] = 1;
+
+        for (auto s : seq) {
+            // TODO(ragnar): Figure out a nice way to deal with uncertain reads.
+            if (s < 0 || s >= alphabet_size)
+                continue;
+            for (int i = sequence_len - 1; i >= 0; --i)
+                for (size_t j = 0; j < ts[i].size(); ++j)
+                    ts[i + 1][alphabet_size * j + s] += ts[i][j];
+        }
+
+        if (normalize) {
+            double nchooset = 1;
+            for (int i = 0; i < sequence_len; ++i)
+                nchooset = nchooset * (seq.size() - i) / (i + 1);
```

Comment:
make nchooset `uint64_t` so that you avoid double operations. The algorithm will work fine, as all divisions have no remainder.

Reply:
`n=10^5`, `t=10` will make this have value roughly `10^43`, which wouldn't fit in an unsigned int64. I prefer doubles everywhere to be safe.